File: engine/compile.js
// COMPILE — deterministic assembly of the dossier + handoff prompt. No model calls.
// The factoid DB and Billy's template stay separate until this moment; the prompt is a render, not a bake.
"use strict";

const fs = require("fs");
const path = require("path");
const S = require("./state");

function loadTemplate() {
  return fs.readFileSync(path.join(__dirname, "..", "templates", "handoff.md"), "utf8");
}

function section(items, empty) {
  return items.length ? items.map((s) => `- ${s}`).join("\n") : `- ${empty}`;
}

function renderHandoff(state, template) {
  const tpl = template || loadTemplate();
  const sorted = S.sortedFactoids(state);
  const ident = S.byCategory(state, "identity").map((f) => f.text);
  const job = S.byCategory(state, "job").map((f) => f.text);
  const sound = S.byCategory(state, "sound").map((f) => f.text);
  // Defense in depth: only short phrases render as sacred, whatever the state holds.
  const sacred = S.byCategory(state, "sacred")
    .map((f) => f.verbatim || f.text)
    .filter((t) => t.split(/\s+/).length <= 12)
    .map((t) => `"${t}"`);
  const bounds = S.byCategory(state, "boundary").map((f) => f.text);
  const phonetic = state.factoids.filter((f) => f.flags.includes("phonetic")).map((f) => f.text);

  const material = sorted
    .filter((f) => ["specific", "scene", "emotion"].includes(f.category))
    .map((f, i) => `${i + 1}. [${f.category}, heat ${f.weight}] ${f.text}${f.flags.includes("wound") ? "  (a wound — customer consented; hold it with care)" : ""}`)
    .join("\n");

  // The five facts the song cannot skip: highest-ranked story material with a real
  // image in it (fragments carry nothing to ground a lyric on).
  const mustAppear = sorted
    .filter((f) => ["specific", "scene"].includes(f.category) && f.text.split(/\s+/).length >= 5)
    .slice(0, 5)
    .map((f, i) => `${i + 1}. ${f.text}`)
    .join("\n");

  const cov = S.coverage(state);
  const thin = Object.entries(cov)
    .filter(([k, v]) => v.need > 0 && v.have < v.need)
    .map(([k, v]) => `${v.label}: THIN — the customer gave little here. Do not invent depth to fill it.`);

  const transcript = state.transcript
    .map((t) => `Q${t.turn}: ${t.q}\nA${t.turn}: ${t.a}`)
    .join("\n\n");

  return tpl
    .replace("{{IDENTITY}}", section(ident.concat(phonetic), "not fully established — see transcript"))
    .replace("{{JOB}}", section(job, "not stated — a warm, giftable tone is safe"))
    .replace("{{FACTOIDS}}", material || "- (thin — see transcript)")
    .replace("{{MUSTAPPEAR}}", mustAppear || "- (thin — ground in whatever the transcript offers)")
    .replace("{{SACRED}}", section(sacred, "none given"))
    .replace("{{BOUNDARIES}}", section(bounds, "none given"))
    .replace("{{SOUND}}", section(sound, "not stated — choose to fit the emotional center"))
    .replace("{{THIN}}", thin.length ? thin.map((s) => `- ${s}`).join("\n") : "- none; all core slots earned")
    .replace("{{TRANSCRIPT}}", transcript);
}

// Operator dossier — everything, structured, conclusions beside their evidence.
function renderDossier(state, meta = {}) {
  const cov = S.coverage(state);
  const lines = [];
  lines.push(`# DOSSIER — ${state.name || "unnamed subject"}`);
  lines.push("");
  lines.push("## Fulfillment");
  lines.push(`- Customer: ${meta.customer || "(from link params)"}`);
  lines.push(`- Channel/order: ${meta.order || "(from link params)"}`);
  lines.push(`- Occasion date: ${meta.date || "(none given)"}`);
  lines.push("");
  lines.push("## Coverage");
  for (const [k, v] of Object.entries(cov)) {
    const mark = v.need === 0 ? (v.have ? "✓" : "—") : v.ok ? "✓" : "THIN";
    lines.push(`- ${mark} ${v.label} (${v.have}${v.need ? "/" + v.need : ""})`);
  }
  lines.push("");
  lines.push("## Factoids (sorted, each with its source)");
  for (const f of S.sortedFactoids(state)) {
    lines.push(`- [${f.category} | heat ${f.weight} | turn ${f.turn}${f.flags.length ? " | " + f.flags.join(",") : ""}] ${f.text}`);
    if (f.verbatim && f.verbatim !== f.text) lines.push(`  > "${f.verbatim}"`);
  }
  lines.push("");
  lines.push("## Transcript (verbatim)");
  for (const t of state.transcript) {
    lines.push(`**Q${t.turn}:** ${t.q}`);
    lines.push(`**A${t.turn}:** ${t.a}`);
    lines.push("");
  }
  return lines.join("\n");
}

// The customer's finish paragraph — their own words mirrored back. Code-rendered, zero generation.
// Screenshot-sized: short punchy quotes only, never paragraphs.
function renderFinish(state) {
  const who = state.name && state.name.length <= 30 ? state.name : "them";
  const quotable = S.sortedFactoids(state)
    .filter((f) => ["specific", "scene", "emotion", "sacred"].includes(f.category))
    .map((f) => {
      // Prefer the shorter of text/verbatim; a good mirror line is 3-20 words.
      const cands = [f.verbatim, f.text].filter(Boolean).map((s) => s.trim())
        .filter((s) => { const w = s.split(/\s+/).length; return w >= 3 && w <= 20; })
        .sort((a, b) => a.length - b.length);
      return cands[0] ? { q: cands[0], w: f.weight } : null;
    })
    .filter(Boolean)
    .slice(0, 4);
  if (!quotable.length) return `Your story is in. Everything you told me about ${who} goes straight to the person writing their song.`;
  return [
    `Here's what I'm carrying to the songwriter — in your words:`,
    ...quotable.map((b) => `“${b.q}”`),
    `That's ${who}. Nobody else. Your story is in — the song is next.`,
  ].join("\n");
}

module.exports = { renderHandoff, renderDossier, renderFinish, loadTemplate };

File: android/app/src/main/java/com/sensorysymphony/storymode/Compile.kt
package com.sensorysymphony.storymode

/** Mirrors engine/compile.js — deterministic assembly. No model calls. */
object Compile {

    private fun section(items: List<String>, empty: String) =
        if (items.isNotEmpty()) items.joinToString("\n") { "- $it" } else "- $empty"

    fun renderHandoff(state: StoryState, template: String): String {
        val sorted = state.sortedFactoids()
        val ident = state.byCategory("identity").map { it.text }
        val job = state.byCategory("job").map { it.text }
        val sound = state.byCategory("sound").map { it.text }
        val sacred = state.byCategory("sacred").map { "\"${it.verbatim.ifBlank { it.text }}\"" }
        val bounds = state.byCategory("boundary").map { it.text }
        val phonetic = state.factoids.filter { "phonetic" in it.flags }.map { it.text }

        val material = sorted.filter { it.category in listOf("specific", "scene", "emotion") }
            .mapIndexed { i, f -> "${i + 1}. [${f.category}, heat ${f.weight}] ${f.text}${if ("wound" in f.flags) "  (a wound — customer consented; hold it with care)" else ""}" }
            .joinToString("\n")

        val thin = state.coverage().entries
            .filter { it.value.second > 0 && it.value.first < it.value.second }
            .map { "${CATEGORIES[it.key]?.label}: THIN — the customer gave little here. Do not invent depth to fill it." }

        val transcript = state.transcript.joinToString("\n\n") { "Q${it.turn}: ${it.q}\nA${it.turn}: ${it.a}" }

        return template
            .replace("{{IDENTITY}}", section(ident + phonetic, "not fully established — see transcript"))
            .replace("{{JOB}}", section(job, "not stated — a warm, giftable tone is safe"))
            .replace("{{FACTOIDS}}", material.ifEmpty { "- (thin — see transcript)" })
            .replace("{{SACRED}}", section(sacred, "none given"))
            .replace("{{BOUNDARIES}}", section(bounds, "none given"))
            .replace("{{SOUND}}", section(sound, "not stated — choose to fit the emotional center"))
            .replace("{{THIN}}", if (thin.isNotEmpty()) thin.joinToString("\n") { "- $it" } else "- none; all core slots earned")
            .replace("{{TRANSCRIPT}}", transcript)
    }

    fun renderDossier(state: StoryState): String {
        val lines = mutableListOf<String>()
        lines.add("# DOSSIER — ${state.name ?: "unnamed subject"}")
        lines.add("")
        lines.add("## Coverage")
        for ((k, v) in state.coverage()) {
            val (have, need, ok) = v
            val mark = if (need == 0) (if (have > 0) "✓" else "—") else if (ok) "✓" else "THIN"
            lines.add("- $mark ${CATEGORIES[k]?.label} ($have${if (need > 0) "/$need" else ""})")
        }
        lines.add("")
        lines.add("## Factoids (sorted, each with its source)")
        for (f in state.sortedFactoids()) {
            lines.add("- [${f.category} | heat ${f.weight} | turn ${f.turn}${if (f.flags.isNotEmpty()) " | " + f.flags.joinToString(",") else ""}] ${f.text}")
            if (f.verbatim.isNotBlank() && f.verbatim != f.text) lines.add("  > \"${f.verbatim}\"")
        }
        lines.add("")
        lines.add("## Transcript (verbatim)")
        for (t in state.transcript) {
            lines.add("**Q${t.turn}:** ${t.q}")
            lines.add("**A${t.turn}:** ${t.a}")
            lines.add("")
        }
        return lines.joinToString("\n")
    }

    /** Screenshot-sized finish: the customer's own words mirrored back. Zero generation. */
    fun renderFinish(state: StoryState): String {
        val who = state.name?.takeIf { it.length <= 30 } ?: "them"
        val quotable = state.sortedFactoids()
            .filter { it.category in listOf("specific", "scene", "emotion", "sacred") }
            .mapNotNull { f ->
                listOf(f.verbatim, f.text).filter { it.isNotBlank() }.map { it.trim() }
                    .filter { it.split(Regex("\\s+")).size in 3..20 }
                    .minByOrNull { it.length }
            }
            .take(4)
        if (quotable.isEmpty()) return "Your story is in. Everything you told me about $who goes straight to the person writing their song."
        return buildString {
            append("Here's what I'm carrying to the songwriter — in your words:\n")
            quotable.forEach { append("“").append(it).append("”\n") }
            append("That's $who. Nobody else. Your story is in — the song is next.")
        }
    }
}
